src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleasePermissionService.cs

```diff
@@ -0,0 +1,183 @@
+using GovUk.Education.ExploreEducationStatistics.Content.Model;
+using GovUk.Education.ExploreEducationStatistics.Content.Model.Database;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.EntityFrameworkCore;
+using static GovUk.Education.ExploreEducationStatistics.Content.Model.ReleaseRole;
+
+namespace GovUk.Education.ExploreEducationStatistics.Admin.Services
+{
+    public class ReleasePermissionService : IReleasePermissionService
+    {
+        private readonly ContentDbContext _contentDbContext;
+        private readonly IPersistenceHelper<ContentDbContext> _persistenceHelper;
+        private readonly IUserReleaseRoleRepository _userReleaseRoleRepository;
+        private readonly IUserService _userService;
+
+        public ReleasePermissionService(
+            ContentDbContext contentDbContext,
+            IPersistenceHelper<ContentDbContext> persistenceHelper,
+            IUserReleaseRoleRepository userReleaseRoleRepository,
+            IUserService userService)
+        {
+            _contentDbContext = contentDbContext;
+            _persistenceHelper = persistenceHelper;
+            _userReleaseRoleRepository = userReleaseRoleRepository;
+            _userService = userService;
+        }
+
+        public async Task<Either<ActionResult, List<ContributorViewModel>>>
+            ListReleaseContributors(Guid publicationId, Guid releaseId)
+        {
+            return await _persistenceHelper
+                .CheckEntityExists<Publication>(publicationId,
+                    query =>
+                        query.Include(r => r.Releases))
+                .OnSuccessDo(publication => _userService
+                    .CheckCanUpdateReleaseRole(publication, Contributor))
+                .OnSuccess(async publication =>
+                {
+                    var release = publication.Releases.SingleOrDefault(r => r.Id == releaseId);
+
+                    if (release == null)
+                    {
+                        return new Either<ActionResult, List<ContributorViewModel>>(new NotFoundResult());
+                    }
+
+                    var allContributorsForRelease = await _contentDbContext.UserReleaseRoles
+                        .Include(urr => urr.User)
+                        .Where(urr =>
+                            urr.ReleaseId == release.Id
+                            && urr.Role == Contributor)
+                        .ToListAsync();
+
+                    var contributorList = allContributorsForRelease
+                        .Select(urr => new ContributorViewModel
+                        {
+                            UserId = urr.UserId,
+                            UserDisplayName = urr.User.DisplayName,
+                            UserEmail = urr.User.Email,
+                        })
+                        .OrderBy(model => model.UserDisplayName)
+                        .ToList();
+
+                    return new Either<ActionResult, List<ContributorViewModel>>(contributorList);
```

Comment:
Don't need to wrap this in an Either, it gets converted to one automatically by Either's implicit converter.

Reply:
This was necessary when I was returning the NotFoundResult to make the compiler not confused